aes/ta/aes_ta.c

```diff
@@ -0,0 +1,432 @@
+		return TEE_SUCCESS;
+	default:
+		EMSG("Invalid mode %u", param);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+}
+
+/*
+ * Process command TA_AES_CMD_PREPARE. See API in aes_ta.h
+ *
+ * Allocate resources required for the ciphering operation.
+ * During ciphering operation, when expect client can:
+ * - update the key materials (provided by client)
+ * - reset the initial vector (provided by client)
+ * - cipher an input buffer into an output buffer (provided by client)
+ */
+static TEE_Result alloc_resources(void *session, uint32_t param_types,
+				  TEE_Param params[4])
+{
+	const uint32_t exp_param_types =
+		TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+				TEE_PARAM_TYPE_VALUE_INPUT,
+				TEE_PARAM_TYPE_VALUE_INPUT,
+				TEE_PARAM_TYPE_NONE);
+	struct aes_cipher *sess;
+	TEE_Attribute attr;
+	TEE_Result res;
+	char *key;
+
+	/* Get ciphering context from session ID */
+	DMSG("Session %p: get ciphering resources", session);
+	sess = (struct aes_cipher *)session;
+
+	/* Safely get the invocation parameters */
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = ta2tee_algo_id(params[0].value.a, &sess->algo);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = ta2tee_key_size(params[1].value.a, &sess->key_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	res = ta2tee_mode_id(params[2].value.a, &sess->mode);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	/*
+	 * Ready to allocate the resources which are:
+	 * - an operation handle, for an AES ciphering of given configuration
+	 * - a transient object that will be use to load the key materials
+	 *   into the AES ciphering operation.
+	 */
+
+	/* Free potential previous operation */
+	if (sess->op_handle != TEE_HANDLE_NULL)
+		TEE_FreeOperation(sess->op_handle);
+
+	/* Allocate operation: AES/CTR, mode and size from params */
+	res = TEE_AllocateOperation(&sess->op_handle,
+				    sess->algo,
+				    sess->mode,
+				    sess->key_size * 8);
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to allocate operation");
+		sess->op_handle = TEE_HANDLE_NULL;
+		goto err;
+	}
+
+	/* Free potential previous transient object */
+	if (sess->key_handle != TEE_HANDLE_NULL)
+		TEE_FreeTransientObject(sess->key_handle);
+
+	/* Allocate transient object according to target key size */
+	res = TEE_AllocateTransientObject(TEE_TYPE_AES,
+					  sess->key_size * 8,
+					  &sess->key_handle);
+	if (res != TEE_SUCCESS) {
+		EMSG("Failed to allocate transient object");
+		sess->key_handle = TEE_HANDLE_NULL;
+		goto err;
+	}
+
+	/*
+	 * When loading a key in the cipher session, set_aes_key()
+	 * will reset the operation and load a key. But we cannot
+	 * reset and operation that has no key yet (GPD TEE Internal
+	 * Core API Specification – Public Release v1.1.1, section
+	 * 6.2.5 TEE_ResetOperation). In consequence, we will load a
+	 * dummy key in the operation so that operation can be reset
+	 * when updating the key.
+	 */
+	key = TEE_Malloc(sess->key_size, 0);
```

Comment:
Is the dummy really necessary? Wouldn't it be enough to set
```c
sess->key_handle = TEE_HANDLE_NULL;
```
and then in `set_aes_key()` check
```c
if (sess->key_handle != TEE_HANDLE_NULL)
    TEE_ResetTransientObject(sess->key_handle);
```
That would seem more natural to me. Am I missing something?